docs/tsp/typeServerProtocol.ts

```diff
@@ -0,0 +1,383 @@
+        GetDiagnostics = 'typeServer/getDiagnostics',
+        // Request to get the version of diagnostics for a specific file.
+        GetDiagnosticsVersion = 'typeServer/getDiagnosticsVersion',
+        // Request to get the type information for a specific node.
+        GetType = 'typeServer/getType',
+        // Request to get the union types for a specific type.
+        GetUnionTypes = 'typeServer/getUnionTypes',
+        // Request to find a member of a type.
+        GetTypeMember = 'typeServer/getTypeMember',
+        // Request to get all overloads of a function or method.
+        GetOverloads = 'typeServer/getOverloads',
+        // Request to get the effective meta class of a type.
+        GetEffectiveMetaClass = 'typeServer/getEffectiveMetaClass',
+        // Request to get the type of a declaration.
+        GetTypeOfDeclaration = 'typeServer/getTypeOfDeclaration',
+        // Request to get symbol declaration information for a node.
+        GetSymbolDeclarationInfo = 'typeServer/getSymbolDeclarationInfo',
+        // Request to get the bound magic method for a type.
+        GetBoundMagicMethod = 'typeServer/getBoundMagicMethod',
+        // Request to resolve an alias declaration.
+        // This request is sent by the client to resolve an alias declaration, which may involve looking up the
+        // declaration in the type server's cache or performing additional resolution logic.
+        ResolveAliasDeclaration = 'typeServer/resolveAliasDeclaration',
+        // Request to resolve an import.
+        ResolveImport = 'typeServer/resolveImport',
+    }
+
+    export enum Notifications {
+        // Notification sent by the server to indicate that the type server has been initialized.
+        Initialized = 'typeServer/initialized',
+        // Notification sent by the server to indicate that the type server has shut down.
+        Shutdown = 'typeServer/shutdown',
+        // Notification sent by the client to indicate that a text document has been opened.
+        TextDocumentOpen = 'textDocument/open',
+        // Notification sent by the client to indicate that a text document has been closed.
+        TextDocumentClose = 'textDocument/close',
+        // Notification sent by the client to indicate that settings have changed.
+        SettingsChange = 'typeServer/settingsChange',
+        // Notification sent by the client to indicate that files have changed.
+        FilesChanged = 'typeServer/filesChanged',
+        // Notification sent by the client to indicate that the type server's cache should be flushed (and snapshots reset)
+        FlushCache = 'typeServer/flushCache',
+    }
+
+    export interface Params {
+        [Requests.Initialize]: { workspace: WorkspaceFolder; initialSettings: string };
+        [Requests.Shutdown]: void;
+        [Requests.GetSnapshot]: void;
+        [Requests.GetDiagnostics]: { uri: string; snapshot: number };
+        [Requests.GetDiagnosticsVersion]: { uri: string };
```

Comment:
Wouldn't `GetDiagnosticsVersion` need to take a snapshot ID as well? If not, how will the caller know whether the version is still valid once the response arrives?

Reply:
I had it here originally but it seemed redundant. Well because there's only ever one version, the most recent one. But maybe that's an implementation detail. 

If a type server did support multiple snapshots for some reason, you might want the version of an older snapshot. I'll add the snapshot back.

Reply:
> only ever one version, the most recent one

That works if the interface is synchronous because you can ask for the diagnostic version immediately after asking for the diagnostics and be assured that it hasn't changed in the meantime. But the TSP interface is asynchronous, so that assumption no longer holds.